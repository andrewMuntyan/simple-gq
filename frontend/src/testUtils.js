import { act } from 'react-test-renderer';
import wait from 'waait';

export const fakeWord = () => ({
  content: 'word',
  id: 'uniqueId',
  createdAt: '2019-09-11T21:13:39.695Z'
});

export const getFakeCategoriesData = () => ({
  categories: [
    {
      id: 'uuid1',
      name: 'category 1',
      __typename: 'Category'
    },
    {
      id: 'uuid2',
      name: 'category 2',
      __typename: 'Category'
    },
    {
      id: 'uuid3',
      name: 'category 3',
      __typename: 'Category'
    }
  ]
});

export const fakeCategory = () => getFakeCategoriesData().categories[0];

export const fakePaginationResponse = (count = 0) => ({
  data: {
    wordsConnection: {
      aggregate: { count, __typename: 'AggregateWord' },
      __typename: 'WordConnection'
    }
  }
});

// see https://github.com/airbnb/enzyme/issues/2073

// Use this in your test after mounting if you need just need to let the query finish without updating the wrapper
export async function actWait(amount = 0) {
  await act(async () => {
    await wait(amount);
  });
}

// Use this in your test after mounting if you want the query to finish and update the wrapper
export async function updateWrapper(wrapper, amount = 0) {
  await act(async () => {
    await wait(amount);
    wrapper.update();
  });
}
