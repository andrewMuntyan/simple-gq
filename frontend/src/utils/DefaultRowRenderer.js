import React from 'react';
import PropTypes from 'prop-types';

const DefaultRowRenderer = ({ name }) => {
  return <div>{name}</div>;
};

DefaultRowRenderer.propTypes = {
  name: PropTypes.string.isRequired
};

export default DefaultRowRenderer;
