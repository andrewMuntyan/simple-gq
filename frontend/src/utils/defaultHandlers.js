import wait from 'waait';

export const defaultSubmit = e => {
  e.preventDefault();
  // eslint-disable-next-line no-console
  console.warn('Default Submit handler is triggered');
  return wait(5000);
};

export const defaultOnSMTH = async () => {
  // eslint-disable-next-line no-console
  console.warn('Default Change handler is triggered');
  return wait(5000);
};

export const noop = () => {};
