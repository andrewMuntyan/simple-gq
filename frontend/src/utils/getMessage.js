export default data => {
  return data instanceof Error
    ? { text: data.toString(), variant: 'error' }
    : { text: 'Done!', variant: 'success' };
};
