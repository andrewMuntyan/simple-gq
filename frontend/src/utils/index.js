export { default as DefaultRowRenderer } from './DefaultRowRenderer';
