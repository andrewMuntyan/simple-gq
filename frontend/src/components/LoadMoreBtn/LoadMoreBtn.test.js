import React from 'react';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { defaultOnSMTH } from '../../utils';
import { waitAndUpdateWrapper } from '../../testUtils';

import { LoadMoreBtn } from '.';

describe('<LoadMoreBtn />', () => {
  it('renders and displays properly', async () => {
    const wrapper = mount(
      <LoadMoreBtn displayedItemsCount={1} allItemsCount={2} />
    );

    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('has correct default props', () => {
    const wrapper = mount(<LoadMoreBtn />);

    const { allItemsCount, displayedItemsCount, onClick } = wrapper.props();
    expect(displayedItemsCount).toEqual(0);
    expect(allItemsCount).toEqual(0);
    expect(onClick).toEqual(defaultOnSMTH);
  });

  it('is hidden with default props', () => {
    const wrapper = mount(<LoadMoreBtn />);
    expect(wrapper.html()).toEqual(null);
  });

  it('sets correct visible state', async () => {
    // not visible
    const wrapper = mount(<LoadMoreBtn />);
    expect(wrapper.html()).toBeFalsy();

    // visible
    wrapper.setProps({
      displayedItemsCount: 1,
      allItemsCount: 2
    });
    await waitAndUpdateWrapper(wrapper);
    expect(wrapper.html()).toBeTruthy();

    // not visible
    wrapper.setProps({
      displayedItemsCount: 2,
      allItemsCount: 1
    });
    await waitAndUpdateWrapper(wrapper);
    expect(wrapper.html()).toBeFalsy();
  });

  it('should call onClick callback and change loading state', async () => {
    const onClick = jest.fn(async () => {
      await wait(100);
    });

    const wrapper = mount(
      <LoadMoreBtn
        displayedItemsCount={1}
        allItemsCount={2}
        onClick={onClick}
      />
    );
    const getBtn = () => wrapper.find('button');
    const getSpinner = () => wrapper.find('[data-test="spinner"]');

    const btn = getBtn();

    // there is button and no spinner
    // btn is not disabled
    expect(btn.exists()).toBeTruthy();
    expect(btn.props().disabled).toBeFalsy();
    expect(getSpinner().exists()).toBeFalsy();

    // click
    btn.simulate('click');

    // loading spinner dispalyed
    // btn is disabled
    await waitAndUpdateWrapper(wrapper);
    const displayedSpinner = getSpinner();
    const disabledBtn = getBtn();
    expect(displayedSpinner.exists()).toBeTruthy();
    expect(disabledBtn.props().disabled).toBeTruthy();

    // loading spinner is hidden
    // btn is enabled
    await waitAndUpdateWrapper(wrapper, 1000);
    const enabledBtn = getBtn();
    const hiddenSpinner = getSpinner();
    expect(enabledBtn.props().disabled).toBeFalsy();
    expect(hiddenSpinner.exists()).toBeFalsy();
  });
});
