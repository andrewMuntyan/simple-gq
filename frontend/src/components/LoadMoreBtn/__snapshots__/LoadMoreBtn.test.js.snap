// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoadMoreBtn /> renders and displays properly 1`] = `
<LoadMoreBtn
  allItemsCount={2}
  displayedItemsCount={1}
  onClick={[Function]}
>
  <WithStyles(ForwardRef(Fab))
    className="makeStyles-loadMore-1"
    color="secondary"
    disabled={false}
    onClick={[Function]}
    variant="extended"
  >
    <ForwardRef(Fab)
      className="makeStyles-loadMore-1"
      classes={
        Object {
          "colorInherit": "MuiFab-colorInherit",
          "disabled": "Mui-disabled",
          "extended": "MuiFab-extended",
          "focusVisible": "Mui-focusVisible",
          "label": "MuiFab-label",
          "primary": "MuiFab-primary",
          "root": "MuiFab-root",
          "secondary": "MuiFab-secondary",
          "sizeMedium": "MuiFab-sizeMedium",
          "sizeSmall": "MuiFab-sizeSmall",
        }
      }
      color="secondary"
      disabled={false}
      onClick={[Function]}
      variant="extended"
    >
      <WithStyles(ForwardRef(ButtonBase))
        className="MuiFab-root makeStyles-loadMore-1 MuiFab-extended MuiFab-secondary"
        component="button"
        disabled={false}
        focusRipple={true}
        focusVisibleClassName="Mui-focusVisible"
        onClick={[Function]}
      >
        <ForwardRef(ButtonBase)
          className="MuiFab-root makeStyles-loadMore-1 MuiFab-extended MuiFab-secondary"
          classes={
            Object {
              "disabled": "Mui-disabled",
              "focusVisible": "Mui-focusVisible",
              "root": "MuiButtonBase-root",
            }
          }
          component="button"
          disabled={false}
          focusRipple={true}
          focusVisibleClassName="Mui-focusVisible"
          onClick={[Function]}
        >
          <button
            className="MuiButtonBase-root MuiFab-root makeStyles-loadMore-1 MuiFab-extended MuiFab-secondary"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiFab-label"
            >
              Load More
              <ForwardRef>
                <WithStyles(ForwardRef(SvgIcon))>
                  <ForwardRef(SvgIcon)
                    classes={
                      Object {
                        "colorAction": "MuiSvgIcon-colorAction",
                        "colorDisabled": "MuiSvgIcon-colorDisabled",
                        "colorError": "MuiSvgIcon-colorError",
                        "colorPrimary": "MuiSvgIcon-colorPrimary",
                        "colorSecondary": "MuiSvgIcon-colorSecondary",
                        "fontSizeInherit": "MuiSvgIcon-fontSizeInherit",
                        "fontSizeLarge": "MuiSvgIcon-fontSizeLarge",
                        "fontSizeSmall": "MuiSvgIcon-fontSizeSmall",
                        "root": "MuiSvgIcon-root",
                      }
                    }
                  >
                    <svg
                      aria-hidden="true"
                      className="MuiSvgIcon-root"
                      focusable="false"
                      role="presentation"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 9h-4V3H9v6H5l7 7 7-7zM5 18v2h14v-2H5z"
                      />
                    </svg>
                  </ForwardRef(SvgIcon)>
                </WithStyles(ForwardRef(SvgIcon))>
              </ForwardRef>
            </span>
            <NoSsr>
              <WithStyles(undefined)
                center={false}
              >
                <ForwardRef(TouchRipple)
                  center={false}
                  classes={
                    Object {
                      "child": "MuiTouchRipple-child",
                      "childLeaving": "MuiTouchRipple-childLeaving",
                      "childPulsate": "MuiTouchRipple-childPulsate",
                      "ripple": "MuiTouchRipple-ripple",
                      "ripplePulsate": "MuiTouchRipple-ripplePulsate",
                      "rippleVisible": "MuiTouchRipple-rippleVisible",
                      "root": "MuiTouchRipple-root",
                    }
                  }
                >
                  <span
                    className="MuiTouchRipple-root"
                  >
                    <TransitionGroup
                      childFactory={[Function]}
                      component={null}
                      exit={true}
                    />
                  </span>
                </ForwardRef(TouchRipple)>
              </WithStyles(undefined)>
            </NoSsr>
          </button>
        </ForwardRef(ButtonBase)>
      </WithStyles(ForwardRef(ButtonBase))>
    </ForwardRef(Fab)>
  </WithStyles(ForwardRef(Fab))>
</LoadMoreBtn>
`;
